### PR TITLE
Matmul, bias epilogue support

### PR DESCRIPTION
**Pull request by drzejan2**
The core functionality introduced with this change:

- update matmul scheduler to recognize bias epilogue pattern, if detected then handled as TensorView with `INPUT_C` role,
- simplify the handling of non-core roles in matmul scheduler,
- add unit test for bias epilogue,
- add functional tests,

### CODE REVIEW

**Comment by drzejan2** [comment]
!build

File: test/test_gpu_tensorcore.cpp [hunk 1]
```diff
@@ -4225,6 +4225,177 @@ TEST_F(NVFuserTest, FusionMatmulSchedulerEpilogueAlphaBetaGeluOutputCast_CUDA) {
   TORCH_CHECK(outputs[0].allclose(t8, 0.01, 0.06));
 }
 
+// Matmul test that relies on segmenter for fusion for Ampere:
+//  D = (A x B) + bias
+TEST_F(NVFuserTest, FusionMatmulSchedulerEpilogueBias_CUDA) {
```
Comment (drzejan2) [inline test/test_gpu_tensorcore.cpp:4230]: The generated CUDA code looks like this for this test case:
```
__global__ void kernel1(Tensor<__half, 2, 2> T0, Tensor<__half, 2, 2> T1, Tensor<float, 1, 1> T2, Tensor<float, 2, 2> T8) {
  alignas(16) extern __shared__ char array[];
  unsigned smem_offset = 0;
  int i0 = T1.size[1] * 8;
  int i1 = ((nvfuser_index_t)threadIdx.x) / 4;
  int i2 = ((nvfuser_index_t)threadIdx.x) % 4;
  int i3 = 8 * i2;
  __half* ptr4 = (T1).data;
  int i5 = T1.size[1] * i1;
  int i6 = i0 * ((nvfuser_index_t)threadIdx.y);
  int i7 = i0 * ((nvfuser_index_t)threadIdx.z);
  int i8 = 32 * ((nvfuser_index_t)blockIdx.y);
  __half* ptr9 = ((((ptr4 + i5) + i3) + i6) + i7) + i8;
  int i10 = T1.size[1] * 32;
  int i11 = 64 * i1;
  int i12 = 64 * (i2 / 4);
  int i13 = 16 * (((i1 / 2) % 4) ^ (i2 % 4));
  int i14 = 512 * ((nvfuser_index_t)threadIdx.y);
  int i15 = 512 * ((nvfuser_index_t)threadIdx.z);
  int i16 = T0.size[1] * 8;
  __half* ptr17 = (T0).data;
  int i18 = T0.size[1] * i1;
  int i19 = i16 * ((nvfuser_index_t)threadIdx.y);
  int i20 = i16 * ((nvfuser_index_t)threadIdx.z);
  int i21 = (T0.size[1] * 256) * ((nvfuser_index_t)blockIdx.x);
  __half* ptr22 = ((((ptr17 + i18) + i3) + i19) + i20) + i21;
  int i23 = T0.size[1] * 32;
  int i24 = ((nvfuser_index_t)threadIdx.x) / 8;
  int i25 = 512 * (i24 % 2);
  int i26 = ((nvfuser_index_t)threadIdx.x) % 8;
  int i27 = 64 * i26;
  int i28 = i24 / 2;
  int i29 = 64 * (i28 / 4);
  int i30 = (i26 / 2) % 4;
  int i31 = 16 * (i30 ^ (i28 % 4));
  int i32 = 4096 * ((nvfuser_index_t)threadIdx.z);
  int i33 = ((nvfuser_index_t)threadIdx.x) % 16;
  int i34 = 64 * i33;
  int i35 = ((nvfuser_index_t)threadIdx.x) / 16;
  int i36 = 8 * (i35 % 2);
  int i37 = 64 * ((nvfuser_index_t)threadIdx.y);
  int i38 = i35 / 2;
  int i39 = (i33 / 2) % 4;
  int i40 = T1.size[1] * 64;
  __half* ptr41 = (((((i40 + ptr4) + i5) + i3) + i6) + i7) + i8;
  __half* ptr42 = (((((64 + ptr17) + i18) + i3) + i19) + i20) + i21;
  int i43 = 2 + i28;
  int i44 = 64 * (i43 / 4);
  int i45 = 16 * (i30 ^ (i43 % 4));
  int i46 = 256 * ((nvfuser_index_t)blockIdx.x);
  int i47 = 64 * ((nvfuser_index_t)threadIdx.z);
  int i48 = (i1 + i47) + i46;
  int i49 = 2 * i2;
  int i50 = 32 * ((nvfuser_index_t)threadIdx.y);
  int i51 = ((((i5 + i49) + i50) + (i40 * ((nvfuser_index_t)threadIdx.z))) + i8) + ((T1.size[1] * 256) * ((nvfuser_index_t)blockIdx.x));
  int i52 = T1.size[1] * 16;
  bool b53 = ((7 + i3) + i8) < T1.size[1];
  int i54 = 8 * ((nvfuser_index_t)threadIdx.y);
  int i55 = 8 * ((nvfuser_index_t)threadIdx.z);
  int i56 = (((-T0.size[1]) + i1) + i54) + i55;
  int i57 = (7 - T0.size[1]) + i3;
  int i58 = (-T0.size[0]) + i1;
  int i59 = ((i58 + i54) + i55) + i46;
  int i60 = ((64 + i1) + i54) + i55;
  int i61 = 71 + i3;
  int i62 = (i58 + i47) + i46;
  int i63 = (((1 - T1.size[1]) + i49) + i50) + i8;
  smem_offset = alignBufferSize(smem_offset, 16);
  __half* T11 = reinterpret_cast<__half*>(array + smem_offset);
  smem_offset += (3072 * sizeof(__half));
  smem_offset = alignBufferSize(smem_offset, 16);
  __half* T10 = reinterpret_cast<__half*>(array + smem_offset);
  smem_offset += (24576 * sizeof(__half));
  unsigned i64 = toSmem((T11));
  unsigned i65 = ((((i64 + i11) + i12) + i13) + i14) + i15;
  unsigned i66 = toSmem((T10));
  unsigned i67 = ((((i66 + i11) + i12) + i13) + i14) + i15;
  unsigned i68 = (i66 + i25) + i27;
  unsigned i69 = ((i68 + i29) + i31) + i32;
  unsigned i70 = ((i64 + i34) + i36) + i37;
  unsigned i71 = ((i68 + i44) + i45) + i32;
  unsigned i72 = (((1024 + i64) + i34) + i36) + i37;
  float T6[64];
  #pragma unroll
  for(nvfuser_index_t i73 = 0; i73 < 4; ++i73) {
    int i74 = 16 * i73;
    #pragma unroll
    for(nvfuser_index_t i75 = 0; i75 < 4; ++i75) {
      Ampere::initM16N8K16TN<8>(reinterpret_cast<Array<float,4,4>*>(&T6[(i74 + (2 * i75))]));
    }
  }
  #pragma unroll
  for(nvfuser_index_t i76 = 0; i76 < 2; ++i76) {
    int i77 = 32 * i76;
    __half* ptr78 = ptr22 + i77;
    unsigned i79 = i67 + (16384 * i76);
    int i80 = -i77;
    bool b81 = i57 < i80;
    Ampere::cpAsyncCg<__half, 8>((i65 + (2048 * i76)), (ptr9 + (i10 * i76)), (b53 && (i56 < i80)));
    #pragma unroll
    for(nvfuser_index_t i82 = 0; i82 < 8; ++i82) {
      Ampere::cpAsyncCg<__half, 8>((i79 + (2048 * i82)), (ptr78 + (i23 * i82)), (b81 && (i59 < (-(32 * i82)))));
    }
    Ampere::cpAsyncCommit();
  }
  Ampere::cpAsyncPartialBarrier<1>();
  __syncthreads();
  Array<__half, 64, 8> T12;
  Array<__half, 32, 4> T3;
  #pragma unroll
  for(nvfuser_index_t i83 = 0; i83 < 4; ++i83) {
    Turing::ldMatrix(reinterpret_cast<Array<__half,8,8>*>(&T12[(8 * i83)]), (i69 + (1024 * i83)));
  }
  #pragma unroll
  for(nvfuser_index_t i84 = 0; i84 < 4; ++i84) {
    int i85 = i38 + i84;
    Turing::ldMatrixT(reinterpret_cast<Array<__half,4,4>*>(&T3[(4 * i84)]), ((i70 + (64 * (i85 / 4))) + (16 * (i39 ^ (i85 % 4)))));
  }
  #pragma unroll 1
  for(nvfuser_index_t i86 = 0; i86 < (ceilDiv(T0.size[1], 32)); ++i86) {
    int i87 = (2 + i86) % 3;
    int i88 = 32 * i86;
    __half* ptr89 = ptr42 + i88;
    unsigned i90 = i67 + (16384 * i87);
    int i91 = i86 % 3;
    unsigned i92 = i71 + (16384 * i91);
    unsigned i93 = i72 + (2048 * i91);
    int i94 = (1 + i86) % 3;
    unsigned i95 = i69 + (16384 * i94);
    unsigned i96 = i70 + (2048 * i94);
    bool b97 = (i61 + i88) < T0.size[1];
    Ampere::cpAsyncCg<__half, 8>((i65 + (2048 * i87)), (ptr41 + (i10 * i86)), (b53 && ((i60 + i88) < T0.size[1])));
    #pragma unroll
    for(nvfuser_index_t i82 = 0; i82 < 8; ++i82) {
      Ampere::cpAsyncCg<__half, 8>((i90 + (2048 * i82)), (ptr89 + (i23 * i82)), (b97 && (i59 < (-(32 * i82)))));
    }
    Ampere::cpAsyncCommit();
    #pragma unroll
    for(nvfuser_index_t i83 = 0; i83 < 4; ++i83) {
      Turing::ldMatrix(reinterpret_cast<Array<__half,8,8>*>(&T12[(32 + (8 * i83))]), (i92 + (1024 * i83)));
    }
    __half T4[32];
    #pragma unroll
    for(nvfuser_index_t i98 = 0; i98 < 4; ++i98) {
      int i99 = 8 * i98;
      #pragma unroll
      for(nvfuser_index_t i100 = 0; i100 < 8; ++i100) {
        T4[(i99 + i100)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i98 = 0; i98 < 4; ++i98) {
      int i101 = 8 * i98;
      #pragma unroll
      for(nvfuser_index_t i100 = 0; i100 < 8; ++i100) {
        int i102 = i101 + i100;
        T4[i102] = T12[i102];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i84 = 0; i84 < 4; ++i84) {
      int i103 = i38 + i84;
      Turing::ldMatrixT(reinterpret_cast<Array<__half,4,4>*>(&T3[(16 + (4 * i84))]), ((i93 + (64 * (i103 / 4))) + (16 * (i39 ^ (i103 % 4)))));
    }
    __half T5[16];
    #pragma unroll
    for(nvfuser_index_t i104 = 0; i104 < 4; ++i104) {
      int i105 = 4 * i104;
      #pragma unroll
      for(nvfuser_index_t i106 = 0; i106 < 4; ++i106) {
        T5[(i105 + i106)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i104 = 0; i104 < 4; ++i104) {
      int i107 = 4 * i104;
      #pragma unroll
      for(nvfuser_index_t i106 = 0; i106 < 4; ++i106) {
        int i108 = i107 + i106;
        T5[i108] = T3[i108];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i73 = 0; i73 < 4; ++i73) {
      int i109 = 16 * i73;
      #pragma unroll
      for(nvfuser_index_t i75 = 0; i75 < 4; ++i75) {
        Ampere::M16N8K16TN<8>(
          reinterpret_cast<Array<float,4,4>*>(&T6[(i109 + (2 * i75))]),
          &(reinterpret_cast<Array<__half,8,8>*>(&T4)[i73]),
          &(reinterpret_cast<Array<__half,4,4>*>(&T5)[i75]));
      }
    }
    #pragma unroll
    for(nvfuser_index_t i98 = 0; i98 < 4; ++i98) {
      int i110 = 8 * i98;
      #pragma unroll
      for(nvfuser_index_t i100 = 0; i100 < 8; ++i100) {
        T4[(i110 + i100)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i98 = 0; i98 < 4; ++i98) {
      int i111 = 8 * i98;
      int i112 = 32 + i111;
      #pragma unroll
      for(nvfuser_index_t i100 = 0; i100 < 8; ++i100) {
        T4[(i111 + i100)] = T12[(i112 + i100)];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i104 = 0; i104 < 4; ++i104) {
      int i113 = 4 * i104;
      #pragma unroll
      for(nvfuser_index_t i106 = 0; i106 < 4; ++i106) {
        T5[(i113 + i106)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i104 = 0; i104 < 4; ++i104) {
      int i114 = 4 * i104;
      int i115 = 16 + i114;
      #pragma unroll
      for(nvfuser_index_t i106 = 0; i106 < 4; ++i106) {
        T5[(i114 + i106)] = T3[(i115 + i106)];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i73 = 0; i73 < 4; ++i73) {
      int i116 = 16 * i73;
      #pragma unroll
      for(nvfuser_index_t i75 = 0; i75 < 4; ++i75) {
        Ampere::M16N8K16TN<8>(
          reinterpret_cast<Array<float,4,4>*>(&T6[(i116 + (2 * i75))]),
          &(reinterpret_cast<Array<__half,8,8>*>(&T4)[i73]),
          &(reinterpret_cast<Array<__half,4,4>*>(&T5)[i75]));
      }
    }
    Ampere::cpAsyncPartialBarrier<1>();
    __syncthreads();
    #pragma unroll
    for(nvfuser_index_t i83 = 0; i83 < 4; ++i83) {
      Turing::ldMatrix(reinterpret_cast<Array<__half,8,8>*>(&T12[(8 * i83)]), (i95 + (1024 * i83)));
    }
    #pragma unroll
    for(nvfuser_index_t i84 = 0; i84 < 4; ++i84) {
      int i117 = i38 + i84;
      Turing::ldMatrixT(reinterpret_cast<Array<__half,4,4>*>(&T3[(4 * i84)]), ((i96 + (64 * (i117 / 4))) + (16 * (i39 ^ (i117 % 4)))));
    }
  }
  #pragma unroll
  for(nvfuser_index_t i118 = 0; i118 < 4; ++i118) {
    int i119 = 16 * i118;
    int i120 = i48 + i119;
    int i121 = i51 + (i52 * i118);
    int i122 = -i119;
    #pragma unroll
    for(nvfuser_index_t i123 = 0; i123 < 4; ++i123) {
      int i124 = i119 + (2 * i123);
      int i125 = 8 * i123;
      int i126 = i121 + i125;
      bool b127 = i63 < (-i125);
      #pragma unroll
      for(nvfuser_index_t i128 = 0; i128 < 2; ++i128) {
        int i129 = 8 * i128;
        int i130 = i124 + i129;
        bool b131 = i62 < (i122 - i129);
        float T7[1];
        T7[0] = 0;
        if (b131) {
          T7[0] = T2[(i120 + i129)];
        }
        Array<float, 2, 2> T9;
        #pragma unroll
        for(nvfuser_index_t i132 = 0; i132 < 2; ++i132) {
          T9[i132]
            = T6[(i130 + i132)]
            + T7[0];
        }
        if ((b127 && b131)) {
          loadLocalToGlobal<float, 2, false>( &T8[(i126 + (i0 * i128))], &T9[0]);
        }
      }
    }
  }
}
```

File: test/test_gpu_tensorcore.cpp [hunk 2]
```diff
@@ -4225,6 +4225,177 @@ TEST_F(NVFuserTest, FusionMatmulSchedulerEpilogueAlphaBetaGeluOutputCast_CUDA) {
...
+      "segmentation did happen");
+
+  // NOTE: increasted absolute tolerance to silence false negative verification
+  //       caused by different way of calculating reference
+  TORCH_CHECK(outputs[0].allclose(t4, 0.001, 0.001));
+}
+
+// Matmul test that relies on segmenter for fusion for Ampere:
+//  D = alpha * ((A x B) + bias) + beta * C
+TEST_F(NVFuserTest, FusionMatmulSchedulerEpilogueAlphaBetaBias_CUDA) {
```
Comment (drzejan2) [inline test/test_gpu_tensorcore.cpp:4307]: The generated CUDA code for this test case looks like this:
```
__global__ void kernel1(Tensor<__half, 2, 2> T0, Tensor<__half, 2, 2> T1, Tensor<__half, 2, 2> T2, Tensor<float, 1, 1> T3, double d0, double d1, Tensor<float, 2, 2> T13) {
  alignas(16) extern __shared__ char array[];
  unsigned smem_offset = 0;
  int i2 = T1.size[1] * 8;
  int i3 = ((nvfuser_index_t)threadIdx.x) / 4;
  int i4 = ((nvfuser_index_t)threadIdx.x) % 4;
  int i5 = 8 * i4;
  __half* ptr6 = (T1).data;
  int i7 = T1.size[1] * i3;
  int i8 = i2 * ((nvfuser_index_t)threadIdx.y);
  int i9 = i2 * ((nvfuser_index_t)threadIdx.z);
  int i10 = 32 * ((nvfuser_index_t)blockIdx.y);
  __half* ptr11 = ((((ptr6 + i7) + i5) + i8) + i9) + i10;
  int i12 = T1.size[1] * 32;
  int i13 = 64 * i3;
  int i14 = 64 * (i4 / 4);
  int i15 = 16 * (((i3 / 2) % 4) ^ (i4 % 4));
  int i16 = 512 * ((nvfuser_index_t)threadIdx.y);
  int i17 = 512 * ((nvfuser_index_t)threadIdx.z);
  int i18 = T0.size[1] * 8;
  __half* ptr19 = (T0).data;
  int i20 = T0.size[1] * i3;
  int i21 = i18 * ((nvfuser_index_t)threadIdx.y);
  int i22 = i18 * ((nvfuser_index_t)threadIdx.z);
  int i23 = (T0.size[1] * 256) * ((nvfuser_index_t)blockIdx.x);
  __half* ptr24 = ((((ptr19 + i20) + i5) + i21) + i22) + i23;
  int i25 = T0.size[1] * 32;
  int i26 = ((nvfuser_index_t)threadIdx.x) / 8;
  int i27 = 512 * (i26 % 2);
  int i28 = ((nvfuser_index_t)threadIdx.x) % 8;
  int i29 = 64 * i28;
  int i30 = i26 / 2;
  int i31 = 64 * (i30 / 4);
  int i32 = (i28 / 2) % 4;
  int i33 = 16 * (i32 ^ (i30 % 4));
  int i34 = 4096 * ((nvfuser_index_t)threadIdx.z);
  int i35 = ((nvfuser_index_t)threadIdx.x) % 16;
  int i36 = 64 * i35;
  int i37 = ((nvfuser_index_t)threadIdx.x) / 16;
  int i38 = 8 * (i37 % 2);
  int i39 = 64 * ((nvfuser_index_t)threadIdx.y);
  int i40 = i37 / 2;
  int i41 = (i35 / 2) % 4;
  int i42 = T1.size[1] * 64;
  __half* ptr43 = (((((i42 + ptr6) + i7) + i5) + i8) + i9) + i10;
  __half* ptr44 = (((((64 + ptr19) + i20) + i5) + i21) + i22) + i23;
  int i45 = 2 + i30;
  int i46 = 64 * (i45 / 4);
  int i47 = 16 * (i32 ^ (i45 % 4));
  int i48 = 2 * i4;
  int i49 = 32 * ((nvfuser_index_t)threadIdx.y);
  int i50 = ((((i7 + i48) + i49) + (i42 * ((nvfuser_index_t)threadIdx.z))) + i10) + ((T1.size[1] * 256) * ((nvfuser_index_t)blockIdx.x));
  int i51 = T1.size[1] * 16;
  int i52 = 256 * ((nvfuser_index_t)blockIdx.x);
  int i53 = 64 * ((nvfuser_index_t)threadIdx.z);
  int i54 = (i3 + i53) + i52;
  bool b55 = ((7 + i5) + i10) < T1.size[1];
  int i56 = 8 * ((nvfuser_index_t)threadIdx.y);
  int i57 = 8 * ((nvfuser_index_t)threadIdx.z);
  int i58 = (((-T0.size[1]) + i3) + i56) + i57;
  int i59 = (7 - T0.size[1]) + i5;
  int i60 = (-T0.size[0]) + i3;
  int i61 = ((i60 + i56) + i57) + i52;
  int i62 = ((64 + i3) + i56) + i57;
  int i63 = 71 + i5;
  int i64 = (((1 - T1.size[1]) + i48) + i49) + i10;
  int i65 = (i60 + i53) + i52;
  smem_offset = alignBufferSize(smem_offset, 16);
  __half* T16 = reinterpret_cast<__half*>(array + smem_offset);
  smem_offset += (3072 * sizeof(__half));
  smem_offset = alignBufferSize(smem_offset, 16);
  __half* T15 = reinterpret_cast<__half*>(array + smem_offset);
  smem_offset += (24576 * sizeof(__half));
  unsigned i66 = toSmem((T16));
  unsigned i67 = ((((i66 + i13) + i14) + i15) + i16) + i17;
  unsigned i68 = toSmem((T15));
  unsigned i69 = ((((i68 + i13) + i14) + i15) + i16) + i17;
  unsigned i70 = (i68 + i27) + i29;
  unsigned i71 = ((i70 + i31) + i33) + i34;
  unsigned i72 = ((i66 + i36) + i38) + i39;
  unsigned i73 = ((i70 + i46) + i47) + i34;
  unsigned i74 = (((1024 + i66) + i36) + i38) + i39;
  float T7[64];
  #pragma unroll
  for(nvfuser_index_t i75 = 0; i75 < 4; ++i75) {
    int i76 = 16 * i75;
    #pragma unroll
    for(nvfuser_index_t i77 = 0; i77 < 4; ++i77) {
      Ampere::initM16N8K16TN<8>(reinterpret_cast<Array<float,4,4>*>(&T7[(i76 + (2 * i77))]));
    }
  }
  #pragma unroll
  for(nvfuser_index_t i78 = 0; i78 < 2; ++i78) {
    int i79 = 32 * i78;
    __half* ptr80 = ptr24 + i79;
    unsigned i81 = i69 + (16384 * i78);
    int i82 = -i79;
    bool b83 = i59 < i82;
    Ampere::cpAsyncCg<__half, 8>((i67 + (2048 * i78)), (ptr11 + (i12 * i78)), (b55 && (i58 < i82)));
    #pragma unroll
    for(nvfuser_index_t i84 = 0; i84 < 8; ++i84) {
      Ampere::cpAsyncCg<__half, 8>((i81 + (2048 * i84)), (ptr80 + (i25 * i84)), (b83 && (i61 < (-(32 * i84)))));
    }
    Ampere::cpAsyncCommit();
  }
  Ampere::cpAsyncPartialBarrier<1>();
  __syncthreads();
  Array<__half, 64, 8> T17;
  Array<__half, 32, 4> T4;
  #pragma unroll
  for(nvfuser_index_t i85 = 0; i85 < 4; ++i85) {
    Turing::ldMatrix(reinterpret_cast<Array<__half,8,8>*>(&T17[(8 * i85)]), (i71 + (1024 * i85)));
  }
  #pragma unroll
  for(nvfuser_index_t i86 = 0; i86 < 4; ++i86) {
    int i87 = i40 + i86;
    Turing::ldMatrixT(reinterpret_cast<Array<__half,4,4>*>(&T4[(4 * i86)]), ((i72 + (64 * (i87 / 4))) + (16 * (i41 ^ (i87 % 4)))));
  }
  #pragma unroll 1
  for(nvfuser_index_t i88 = 0; i88 < (ceilDiv(T0.size[1], 32)); ++i88) {
    int i89 = (2 + i88) % 3;
    int i90 = 32 * i88;
    __half* ptr91 = ptr44 + i90;
    unsigned i92 = i69 + (16384 * i89);
    int i93 = i88 % 3;
    unsigned i94 = i73 + (16384 * i93);
    unsigned i95 = i74 + (2048 * i93);
    int i96 = (1 + i88) % 3;
    unsigned i97 = i71 + (16384 * i96);
    unsigned i98 = i72 + (2048 * i96);
    bool b99 = (i63 + i90) < T0.size[1];
    Ampere::cpAsyncCg<__half, 8>((i67 + (2048 * i89)), (ptr43 + (i12 * i88)), (b55 && ((i62 + i90) < T0.size[1])));
    #pragma unroll
    for(nvfuser_index_t i84 = 0; i84 < 8; ++i84) {
      Ampere::cpAsyncCg<__half, 8>((i92 + (2048 * i84)), (ptr91 + (i25 * i84)), (b99 && (i61 < (-(32 * i84)))));
    }
    Ampere::cpAsyncCommit();
    #pragma unroll
    for(nvfuser_index_t i85 = 0; i85 < 4; ++i85) {
      Turing::ldMatrix(reinterpret_cast<Array<__half,8,8>*>(&T17[(32 + (8 * i85))]), (i94 + (1024 * i85)));
    }
    __half T5[32];
    #pragma unroll
    for(nvfuser_index_t i100 = 0; i100 < 4; ++i100) {
      int i101 = 8 * i100;
      #pragma unroll
      for(nvfuser_index_t i102 = 0; i102 < 8; ++i102) {
        T5[(i101 + i102)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i100 = 0; i100 < 4; ++i100) {
      int i103 = 8 * i100;
      #pragma unroll
      for(nvfuser_index_t i102 = 0; i102 < 8; ++i102) {
        int i104 = i103 + i102;
        T5[i104] = T17[i104];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i86 = 0; i86 < 4; ++i86) {
      int i105 = i40 + i86;
      Turing::ldMatrixT(reinterpret_cast<Array<__half,4,4>*>(&T4[(16 + (4 * i86))]), ((i95 + (64 * (i105 / 4))) + (16 * (i41 ^ (i105 % 4)))));
    }
    __half T6[16];
    #pragma unroll
    for(nvfuser_index_t i106 = 0; i106 < 4; ++i106) {
      int i107 = 4 * i106;
      #pragma unroll
      for(nvfuser_index_t i108 = 0; i108 < 4; ++i108) {
        T6[(i107 + i108)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i106 = 0; i106 < 4; ++i106) {
      int i109 = 4 * i106;
      #pragma unroll
      for(nvfuser_index_t i108 = 0; i108 < 4; ++i108) {
        int i110 = i109 + i108;
        T6[i110] = T4[i110];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i75 = 0; i75 < 4; ++i75) {
      int i111 = 16 * i75;
      #pragma unroll
      for(nvfuser_index_t i77 = 0; i77 < 4; ++i77) {
        Ampere::M16N8K16TN<8>(
          reinterpret_cast<Array<float,4,4>*>(&T7[(i111 + (2 * i77))]),
          &(reinterpret_cast<Array<__half,8,8>*>(&T5)[i75]),
          &(reinterpret_cast<Array<__half,4,4>*>(&T6)[i77]));
      }
    }
    #pragma unroll
    for(nvfuser_index_t i100 = 0; i100 < 4; ++i100) {
      int i112 = 8 * i100;
      #pragma unroll
      for(nvfuser_index_t i102 = 0; i102 < 8; ++i102) {
        T5[(i112 + i102)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i100 = 0; i100 < 4; ++i100) {
      int i113 = 8 * i100;
      int i114 = 32 + i113;
      #pragma unroll
      for(nvfuser_index_t i102 = 0; i102 < 8; ++i102) {
        T5[(i113 + i102)] = T17[(i114 + i102)];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i106 = 0; i106 < 4; ++i106) {
      int i115 = 4 * i106;
      #pragma unroll
      for(nvfuser_index_t i108 = 0; i108 < 4; ++i108) {
        T6[(i115 + i108)] = 0.00000000000000000e+00;
      }
    }
    #pragma unroll
    for(nvfuser_index_t i106 = 0; i106 < 4; ++i106) {
      int i116 = 4 * i106;
      int i117 = 16 + i116;
      #pragma unroll
      for(nvfuser_index_t i108 = 0; i108 < 4; ++i108) {
        T6[(i116 + i108)] = T4[(i117 + i108)];
      }
    }
    #pragma unroll
    for(nvfuser_index_t i75 = 0; i75 < 4; ++i75) {
      int i118 = 16 * i75;
      #pragma unroll
      for(nvfuser_index_t i77 = 0; i77 < 4; ++i77) {
        Ampere::M16N8K16TN<8>(
          reinterpret_cast<Array<float,4,4>*>(&T7[(i118 + (2 * i77))]),
          &(reinterpret_cast<Array<__half,8,8>*>(&T5)[i75]),
          &(reinterpret_cast<Array<__half,4,4>*>(&T6)[i77]));
      }
    }
    Ampere::cpAsyncPartialBarrier<1>();
    __syncthreads();
    #pragma unroll
    for(nvfuser_index_t i85 = 0; i85 < 4; ++i85) {
      Turing::ldMatrix(reinterpret_cast<Array<__half,8,8>*>(&T17[(8 * i85)]), (i97 + (1024 * i85)));
    }
    #pragma unroll
    for(nvfuser_index_t i86 = 0; i86 < 4; ++i86) {
      int i119 = i40 + i86;
      Turing::ldMatrixT(reinterpret_cast<Array<__half,4,4>*>(&T4[(4 * i86)]), ((i98 + (64 * (i119 / 4))) + (16 * (i41 ^ (i119 % 4)))));
    }
  }
  #pragma unroll
  for(nvfuser_index_t i120 = 0; i120 < 4; ++i120) {
    int i121 = i50 + (i51 * i120);
    int i122 = 16 * i120;
    int i123 = i54 + i122;
    int i124 = -i122;
    #pragma unroll
    for(nvfuser_index_t i125 = 0; i125 < 4; ++i125) {
      int i126 = 8 * i125;
      int i127 = i121 + i126;
      int i128 = i122 + (2 * i125);
      bool b129 = i64 < (-i126);
      #pragma unroll
      for(nvfuser_index_t i130 = 0; i130 < 2; ++i130) {
        int i131 = i127 + (i2 * i130);
        int i132 = 8 * i130;
        int i133 = i128 + i132;
        bool b134 = i65 < (i124 - i132);
        bool b135 = b129 && b134;
        Array<__half, 2, 2> T18;
        T18.set(__half(0));
        if (b135) {
          loadGlobalToLocal<__half, 2, false>(&T18[0],  &T2[i131]);
        }
        float T8[1];
        T8[0] = 0;
        if (b134) {
          T8[0] = T3[(i123 + i132)];
        }
        Array<float, 2, 2> T14;
        #pragma unroll
        for(nvfuser_index_t i136 = 0; i136 < 2; ++i136) {
          float T11[1];
          T11[0] = __half2float(T18[i136]);
          float T12[1];
          T12[0]
            = (float) d1
            * T11[0];
          float T9[1];
          T9[0]
            = T7[(i133 + i136)]
            + T8[0];
          float T10[1];
          T10[0]
            = (float) d0
            * T9[0];
          T14[i136]
            = T10[0]
            + T12[0];
        }
        if (b135) {
          loadLocalToGlobal<float, 2, false>( &T13[i131], &T14[0]);
        }
      }
    }
  }
}
```

**Comment by drzejan2** [comment]
!build

File: csrc/scheduler/matmul.cpp [hunk 3]
```diff
@@ -634,6 +634,13 @@ void scheduleFusionInputsForEpilogue(const mma_utils::RolesMap& roles_map) {
     // The cached INPUT_C tvs are not needed anymore
     cached_tvs.clear();
   }
+
+  if (roles_map.count(MatmulRole::BIAS)) {
```
[the remark this answers is not in the document]
Comment (drzejan2) [reply csrc/scheduler/matmul.cpp:638]: I need to check few things, currently having the same execution path for `INPU_C` and `BIAS` roles won't work due to asserts and failures.

This is also the initial reason why I decided to split these two things, but as you mentioned in the other comment ([link](https://github.com/NVIDIA/Fuser/pull/583#discussion_r1269516244)), in a long run it introduce too much of maintenance cost.

**Comment by drzejan2** [comment]
!build

**Comment by drzejan2** [comment]
!build

File: test/utils.cpp [hunk 4]
```diff
@@ -599,4 +601,51 @@ void validateSegmentation(
...
+  TORCH_CHECK(
+      tensor.size(tensor.dim() - 2) == rows,
+      "Tensor must have the same number of rows as bias vector");
+
+  auto biasExt = tensor.clone();
+  for (int64_t row = 0; row < rows; ++row) {
+    biasExt[row].copy_(bias[row]);
+  }
+
+  return tensor.add(biasExt);
```
Comment (zasdfgbnm) [inline test/utils.cpp:648]: Why not just `return tensor.add(bias);`?

**Review by zasdfgbnm — APPROVED** [review]

File: csrc/scheduler/matmul.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (drzejan2) [reply csrc/scheduler/matmul.cpp]: The latest revision of the PR contains all needed changes. I'm marking this conversation as resolved.

**Pull request merged**